tuned_lens/nn/unembed.py

```diff
@@ -120,14 +87,23 @@ def load_from_model(self, model: PreTrainedModel):
         U -= U.mean(dim=1, keepdim=True)
         self.unembedding.weight.data = U.to(raw_unembed.weight.dtype)
 
-        self.register_buffer("U_pinv", U.pinverse())
+        # We precompute the pseudo-inverse of the unembedding matrix
+        self.register_buffer("_U_pinv", U.pinverse())
 
-        if self.num_transformer_layers:
+        if extra_layers > 0:
             _, layers = get_transformer_layers(model)
             self.transformer_layers.extend(
-                layers[-self.num_transformer_layers :]  # type: ignore[arg-type]
+                layers[-self.config.extra_layers, :]  # type: ignore[arg-type]
             )
 
+        # In general we don't want to finetune the unembed operation.
+        self.requires_grad_(False)
+
+    def unembedding_hash(self) -> int:
+        """Hash the unmbedding matrix to identify the model."""
+        parameter = self.unembedding.weight.data
+        return hash(str(parameter.detach().cpu().numpy()))
```

Comment:
This doesn't work as a general identifier; Python `hash` is nondeterministic across runs!
<img width="683" alt="Captura de pantalla 2023-04-20 a la(s) 6 26 15 p m" src="https://user-images.githubusercontent.com/39116809/233519101-ae322039-5222-47d4-80c3-0eca6ac48406.png">

Hashing the unembedding is cool, but we should use a hash function from `hashlib` like MD5, SHA, or BLAKE. Also if the weights get cast to a different precision this would fail but that's probably ok

Reply:
Okay good catch thank you